### DISCUSSION TITLE
What do i need to consider when using middleware for Authentication checks

### DISCUSSION THREAD

**Discussion by tn02103** [vercel/next.js#69927] [Help]
### Summary

I am writing my Bachelor-thesis on the secure implementation of the new Next.js App structure, and I have some questions reguarding the middleware. 
## How god is the value req.nextUrl.pathname Decoded an Normalized?
I couldn't find any source on how secure this parameter is and if i have to do some special Validation for this variable before using it in my Authroization Checks. 

In the middleware it is possible to use the req.nextUrl.pathname to do specific checks for different Urls. For example to skip the check if the user is authenticated with the following if-statement. 
`  if (pathname.startsWith("/public") || pathname.startsWith("/login")) {
            console.log("unprotected Route");
            return;
        }
`;

In theory if the value isn't decoded and normalized it would be posible to use PathTraversel to trigger theses kinds of if-Statements for other routs e.g. '/public/../admin/dashboard'. I noticed that the value is getting normalized in such cases and even if i try to use URL-Encoded routes like '/public/%2e%2e/admin/dashboard' is getting decoded an normilized, as long as no '/' are encoded. In that case NextJs doesnt decode or normalize the variable, but allso isnt able to find the page anymore.

**Comment by Shriharsh-Deshmukh** [comment]
Hey there! For your thesis on the secure implementation of Next.js, regarding `req.nextUrl.pathname`, here's a quick breakdown:

The `pathname` is decoded and normalized by Next.js in most cases, preventing typical path traversal attacks like `/public/../admin/dashboard`. However, if certain characters like '/' are URL-encoded, Next.js doesn’t decode them fully, so you might lose track of paths but won't bypass security easily.

Using `startsWith` is generally safe, but it’s a good idea to implement extra validation, especially for edge cases with URL-encoded segments. Better to whitelist allowed paths and add custom validations to be sure no one bypasses your checks.

Let me know if you need more tips!
